### PR TITLE
icedtea_web: 1.7.1 -> 1.7.2 (plus CVE patches)

### PR DESCRIPTION
**Pull request by stefano-m**
On Wed, 31 Jul 2019 it was announced that IcedTea-Web was affected by the below
security vulnerabilities:

- CVE-2019-10185: zip-slip attack during auto-extraction of a JAR file.

- CVE-2019-10181: executable code could be injected in a JAR file without
  compromising the signature verification.

- CVE-2019-10182: improper path sanitization from elements in JNLP
  files.

Version 1.7 was patched, but no release was made. Moreover, the patches apply
cleanly only to 1.7.2, not the current 1.7.1.

Rather than marking 1.7.1 as insecure, update to 1.7.2 and apply the official
patches.

~Note that the patch file has been edited to remove the addition of binary
blobs (that cannot be applied).~

References:

https://www.openwall.com/lists/oss-security/2019/07/31/2
https://github.com/AdoptOpenJDK/IcedTea-Web/issues/327
https://github.com/AdoptOpenJDK/IcedTea-Web/pull/346

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Same manual testing as done in #66422 but using the shell launchers instead of the rust ones.

###### Notify maintainers

cc @grahamc (NixOS security team)
@worldofpeace (reviewer of related #66422)

### CODE REVIEW

**Comment by worldofpeace** [comment]
> Note that the patch file has been edited to remove the addition of binary
blobs (that cannot be applied).

I believe this is because `patchPhase` in `setup.sh` uses `patch` and only `git apply` can do binary.
It's kinda undesirable to have really large patches like this in nixpkgs. 
Perhaps we could fetch the patches and in `postPatch` use `git apply`?
You'd have to use a git repo though, fetched with `fetchgit` and using `leaveDotGit`.

Also, is it possible these binary blobs were important?

If doing that proves too difficult, maybe we can upload it somewhere else.

**Comment by stefano-m** [comment]
@worldofpeace the patch comes straight from https://github.com/AdoptOpenJDK/IcedTea-Web/pull/346 (i.e. https://patch-diff.githubusercontent.com/raw/AdoptOpenJDK/IcedTea-Web/pull/346.patch)

The binary blobs are

- `tests/netx/unit/net/sourceforge/jnlp/runtime/j1.jar`
- `tests/netx/unit/net/sourceforge/jnlp/runtime/jar03_dotdotN1.jar`

They are used in tests, the Changelog in the diff has some more information.  I don't think that the nix package has ever run all tests, so I think it should be OK to not have them in the scope of this PR.

Looking forward to hearing from you.

Thanks

**Comment by worldofpeace** [comment]
Ok, 

> @worldofpeace the patch comes straight from [AdoptOpenJDK/IcedTea-Web#346](https://github.com/AdoptOpenJDK/IcedTea-Web/pull/346) (i.e. https://patch-diff.githubusercontent.com/raw/AdoptOpenJDK/IcedTea-Web/pull/346.patch)
> 
> The binary blobs are
> 
>     * `tests/netx/unit/net/sourceforge/jnlp/runtime/j1.jar`
> 
>     * `tests/netx/unit/net/sourceforge/jnlp/runtime/jar03_dotdotN1.jar`
> 
> 
> They are used in tests, the Changelog in the diff has some more information. I don't think that the nix package has ever run all tests, so I think it should be OK to not have them in the scope of this PR.
> 
> Looking forward to hearing from you.
> 
> Thanks

Ok, that clears up "Also, is it possible these binary blobs were important?" but I still think we shouldn't have such a large patch in nixpkgs for practical reasons.

Do you disagree @grahamc?

**Comment by stefano-m** [comment]
> but I still think we shouldn't have such a large patch in nixpkgs for practical reasons.

I'm not sure I understand. I have removed the blobs from the patch, so it shouldn't be too big. It should include just the code changes.

```shell
$ du -sh pkgs/development/compilers/icedtea-web/cve-fixes.patch
48K     pkgs/development/compilers/icedtea-web/cve-fixes.patch
```

Am I missing something?

**Comment by worldofpeace** [comment]
> > but I still think we shouldn't have such a large patch in nixpkgs for practical reasons.
> 
> I'm not sure I understand. I have removed the blobs from the patch, so it shouldn't be too big. It should include just the code changes.
> 
> ```shell
> $ du -sh pkgs/development/compilers/icedtea-web/cve-fixes.patch
> 48K     pkgs/development/compilers/icedtea-web/cve-fixes.patch
> ```
> 
> Am I missing something?

Yes, we have a guideline to try our best to fetch patches from remote places and not keep them in nixpkgs. I personally support this since it keeps nixpkgs much smaller.

**Comment by stefano-m** [comment]
> Yes, we have a guideline to try our best to fetch patches from remote places and not keep them in nixpkgs. I personally support this since it keeps nixpkgs much smaller.

I see! Thanks for the explanation. I'm happy to rework this to include the patch remotely then.

**Comment by stefano-m** [comment]
I removed the patch file and use `fetchurl` instead. Then use `git apply` in `postPatch` as suggested.

The package builds fine locally and the executables continue to work.

**Comment by worldofpeace** [comment]
Can we remove `gtk2` here as well?

**Comment by worldofpeace** [comment]
Can we remove `gtk2` here as well?

File: pkgs/development/compilers/icedtea-web/default.nix [hunk 1]
```diff
@@ -1,17 +1,31 @@
-{ stdenv, fetchurl, jdk, gtk2, xulrunner, zip, pkgconfig, perl, npapi_sdk, bash, bc }:
+{ stdenv, fetchurl, jdk, glib, xulrunner, zip, pkgconfig, perl, npapi_sdk, bash, bc, git }:
+let
+
+cveFixes = fetchurl {
```
Comment (worldofpeace) [inline pkgs/development/compilers/icedtea-web/default.nix]: Can you put something similar to what's in your summary here?
i.e what cve's are fixed?

**Comment by worldofpeace** [comment]
Testing now, this should be good with the above comment.

**Comment by worldofpeace** [comment]
There's a distinct change in executable names

**Before**
```
result/bin
├── itweb-settings
├── javaws
└── policyeditor
```

**After**
```
result/bin
├── itweb-settings.sh
├── itw-modularjdk.args
├── javaws.sh
└── policyeditor.sh
```

I'd say that's backwards incompatible.

File: pkgs/development/compilers/icedtea-web/default.nix (hunk 1 above)
Comment (stefano-m) [reply pkgs/development/compilers/icedtea-web/default.nix]: done

**Comment by stefano-m** [comment]
> There's a distinct change in executable names

@worldofpeace are you sure about that?

I get the following:

```
$ tree result/bin
result/bin
├── itweb-settings.sh
├── itw-modularjdk.args
├── javaws.sh
└── policyeditor.sh
```

The executable names change in version 1.8.x, but should stay the same in 1.7.2

**Comment by worldofpeace** [comment]
Building 65729e08178c25139683316d3c246e35a4643e80 I get
```
result/bin
├── itweb-settings
├── javaws
└── policyeditor
```

On 7a859977bcdc3461bf95129612a7945bb452c8c2
```
result/bin
├── itweb-settings.sh
├── itw-modularjdk.args
├── javaws.sh
└── policyeditor.sh
```

Perhaps something's up with my setup, could you build those two revisions and see the diff?

**Comment by stefano-m** [comment]
I see! My change definitely builds `sh` files. I will have a look at the difference. I guess that we could always remove the `.sh` extension in a post install phase.

Thanks for spotting that!

**Comment by stefano-m** [comment]
I checked and the executables are all shell scripts. I will remove the extension and filter out `itw-modularjdk.args`. I'm not sure why the difference in outputs though.

**Comment by stefano-m** [comment]
Removed `sh` extensions in 97a0d2c

Note that we need to keep `bin/itw-modularjdk.args` because it's referenced in the launchers though.

Re-tested manually and working as expected.

**Review by worldofpeace — APPROVED** [review]
Still works, thanks for fixing this :+1:

**Pull request merged**